src/hotspot/share/cds/aotConstantPoolResolver.cpp

```diff
@@ -318,6 +327,201 @@ void AOTConstantPoolResolver::maybe_resolve_fmi_ref(InstanceKlass* ik, Method* m
+  ResourceMark rm;
+  for (SignatureStream ss(sig); !ss.is_done(); ss.next()) {
+    if (ss.is_reference()) {
+      Symbol* type = ss.as_symbol();
+      Klass* k = find_loaded_class(Thread::current(), cp->pool_holder()->class_loader(), type);
+      if (k == nullptr) {
+        return false;
+      }
+
+      if (SystemDictionaryShared::check_for_exclusion(k)) {
+        if (log_is_enabled(Warning, cds, resolve)) {
+          ResourceMark rm;
+          log_warning(cds, resolve)("Cannot aot-resolve Lambda proxy because %s is excluded", k->external_name());
+        }
+        return false;
+      }
+
+      if (check_return_type && ss.at_return_type()) {
+        // This is the interface type implemented by the lambda proxy
+        if (!k->is_interface()) {
+          // pool_holder doesn't look like a valid class generated by javac
+          return false;
+        }
+
+        if (has_clinit(InstanceKlass::cast(k))) {
+          // We initialize the class of the archived lambda proxy at VM start-up, which will also initialize
+          // the interface that it implements. If that interface has a clinit method, we can potentially
+          // change program execution order.
+          if (log_is_enabled(Debug, cds, resolve)) {
+            ResourceMark rm;
+            log_debug(cds, resolve)("Cannot aot-resolve Lambda proxy of interface type %s (has <cilint>)", k->external_name());
+          }
+          return false;
+        }
+      }
+    }
+  }
+  return true;
+}
+
+bool AOTConstantPoolResolver::check_lambda_metafactory_methodtype_arg(ConstantPool* cp, int bsms_attribute_index, int arg_i) {
+  int mt_index = cp->operand_argument_index_at(bsms_attribute_index, arg_i);
+  if (!cp->tag_at(mt_index).is_method_type()) {
+    // malformed class?
+    return false;
+  }
+
+  Symbol* sig = cp->method_type_signature_at(mt_index);
+  if (log_is_enabled(Debug, cds, resolve)) {
+    ResourceMark rm;
+    log_debug(cds, resolve)("Checking MethodType for LambdaMetafactory BSM arg %d: %s", arg_i, sig->as_C_string());
+  }
+
+  return check_lambda_metafactory_signature(cp, sig, false);
+}
+
+bool AOTConstantPoolResolver::check_lambda_metafactory_methodhandle_arg(ConstantPool* cp, int bsms_attribute_index, int arg_i) {
+  int mh_index = cp->operand_argument_index_at(bsms_attribute_index, arg_i);
+  if (!cp->tag_at(mh_index).is_method_handle()) {
+    // malformed class?
+    return false;
+  }
+
+ Symbol* sig = cp->method_handle_signature_ref_at(mh_index);
+  if (log_is_enabled(Debug, cds, resolve)) {
+    ResourceMark rm;
+    log_debug(cds, resolve)("Checking MethodType of MethodHandle for  LambdaMetafactory BSM arg %d: %s", arg_i, sig->as_C_string());
+  }
+  if (!check_lambda_metafactory_signature(cp, sig, false)) {
+    return false;
+  }
+
+  Symbol* mh_sig = cp->method_handle_signature_ref_at(mh_index);
+  if (!check_lambda_metafactory_signature(cp, mh_sig, false)) {
```

Comment:
This appears to be rerunning the previous check without a log message -- or am I missing something here?

Reply:
It was a cut-and-paste mistake. I've removed this line in [195303f](https://github.com/openjdk/jdk/pull/21143/commits/195303f2c6939b93b6152c6610bd38faff5ae3c7) , where I also fixed some minor bugs with `AOTConstantPoolResolver::is_indy_resolution_deterministic()` and added more comments in the C++ code about what checks are being made.

I also added tests code in AOTLinkedLambdas.java to cover all the type checks that happen inside `is_indy_resolution_deterministic()`.